batch/batch/front_end/query/query.py

```diff
@@ -0,0 +1,321 @@
+import abc
+from enum import Enum
+from typing import Any, List, Tuple
+
+from hailtop.utils import parse_timestamp_msecs
+
+from ...exceptions import QueryError
+from .operators import (
+    ComparisonOperator,
+    ExactMatchOperator,
+    MatchOperator,
+    NotEqualExactMatchOperator,
+    PartialMatchOperator,
+    get_operator,
+)
+
+
+class State(Enum):
+    PENDING = 'Pending'
+    READY = 'Ready'
+    CREATING = 'Creating'
+    RUNNING = 'Running'
+    CANCELLED = 'Cancelled'
+    ERROR = 'Error'
+    FAILED = 'Failed'
+    SUCCESS = 'Success'
+
+
+state_search_term_to_states = {
+    'pending': [State.PENDING],
+    'ready': [State.READY],
+    'creating': [State.CREATING],
+    'running': [State.RUNNING],
+    'live': [State.READY, State.CREATING, State.RUNNING],
+    'cancelled': [State.CANCELLED],
+    'error': [State.ERROR],
+    'failed': [State.FAILED],
+    'bad': [State.ERROR, State.FAILED],
+    'success': [State.SUCCESS],
+    'done': [State.CANCELLED, State.ERROR, State.FAILED, State.SUCCESS],
+}
+
+
+def parse_int(word: str) -> int:
+    try:
+        return int(word)
+    except ValueError as e:
+        raise QueryError(f'expected int, but found {word}') from e
+
+
+def parse_float(word: str) -> float:
+    try:
+        return float(word)
+    except ValueError as e:
+        raise QueryError(f'expected float, but found {word}') from e
+
+
+def parse_date(word: str) -> int:
+    try:
+        return parse_timestamp_msecs(word)
+    except ValueError as e:
+        raise QueryError(f'expected date, but found {word}') from e
+
+
+def parse_cost(word: str) -> float:
+    word = word.lstrip('$')
+    return parse_float(word)
+
+
+class Query(abc.ABC):
+    @abc.abstractmethod
+    def query(self) -> Tuple[str, List[Any]]:
+        raise NotImplementedError
+
+
+class StateQuery(Query):
+    @staticmethod
+    def parse(op: str, state: str) -> 'StateQuery':
+        operator = get_operator(op)
+        if not isinstance(operator, ExactMatchOperator):
+            raise QueryError(f'unexpected operator "{op}" expected {ExactMatchOperator}')
```

Comment:
Curious, what does this look like with the actual class name interpolated into this string?